torch/utils/checkpoint.py

```diff
@@ -0,0 +1,114 @@
+        return outputs
+
+    @staticmethod
+    def backward(ctx, *args):
+        if (not torch.autograd.is_checkpoint_valid()):
+            raise RuntimeError("Checkpointing is not compatible with .grad(), please use .backward() if possible")
+        inputs = ctx.saved_tensors
+        inputs_list = detach_variable(inputs)
+        with torch.enable_grad():
+            outputs = ctx.run_function(*inputs_list)
+
+        if torch.is_tensor(outputs):
+            outputs = (outputs,)
+        torch.autograd.backward(outputs, args)
+
+        input_grads = None
+        if isinstance(inputs_list, tuple):
+            input_grads = tuple(inp.grad for inp in inputs_list)
+            return (None,) + input_grads
+        elif torch.is_tensor(inputs_list):
+            input_grads = inputs_list.grad
+            return None, input_grads
+
+
+def checkpoint(run_function, *args):
+    r"""Checkpoint a model or part of the model
+
+    Checkpoint works by trading compute for memory. It can be applied on any
+    part of the model. In the forward pass, the model activations are not
+    stored. The forward pass save the inputs tuple and the run_function
+    parameter. In the backwards pass, the saved inputs and run_function is
+    retreived, and the forward pass is done on the model again (non-volatile
+    this time) since we need to get the activations values for calculating the
+    gradient and then the gradients are calculated.
+    WARNING: checkpointing doesn't work with torch.autograd.grad(), but only with
+    torch.autograd.backward()
+
+    Args:
+        run_function : describes what to run in the forward pass of the model or
+                       part of the model. It should also know how to handle
+                       the inputs passed as the tuple. For example, in LSTM,
+                       user passes (activation, hidden), run_function should
+                       correctly use first input as activation and second input
+                       as hidden
+        args:         tuple containing inputs to the run_function
+
+    Returns:
+        Output of running the run_function on *args
+    """
+    return CheckpointFunction.apply(run_function, *args)
+
+
+def checkpoint_sequential(modules, segments, *inputs):
+    r"""A helper function for checkpointing sequential based models.
+
+    For models that are constructed using sequential, they normally are built
+    using various modules. For such models, given a list of modules it executes
+    sequentially, we can divide the model in various segments and checkpoint
+    the segments. All segments except the last will be run in volatile manner.
+    The inputs of each checkpointed segment will be saved for re-running the
+    segment in the backward pass.
+
+    Args:
+        modules: The sequence of modules (comprising the model) to run in order.
+        segments: Number of times chunks to create in the model
+        inputs: tuple containing the inputs to run_function
```

Comment:
Restriction: all intermediates should be tuples of tensors. This is not mentioned anywhere, but `save_for_backward` will fail if this rule is violated

Reply:
I didn't quite understand what you mean @apaszke ? can you comment more on what you mean?

Reply:
I mean that with the code written as is, there's an extra restriction on the inputs, which is undocumented.